internal/schemautil/service_user_config.go

```diff
@@ -0,0 +1,98 @@
+package schemautil
+
+import (
+	"reflect"
+)
+
+// fillInServiceUserConfigData prepares API compatible service user_config
+func fillInServiceUserConfigData(uc interface{}, b map[string]interface{}) map[string]interface{} {
+	if uc, ok := uc.([]interface{}); ok {
+		if len(uc) == 0 {
+			return nil
+		}
+
+		if uc[0] == nil {
+			return nil
+		}
+
+		// TODO: add logic for m3db namespaces similar to ip_filter
+		flattenUserConfigIPFilter(uc[0].(map[string]interface{}))
+
+		return walkBlueprint(uc[0].(map[string]interface{}), b)
+	}
+	return nil
+}
+
+// walkBlueprint is a function that recursively traverses two maps c and b
+// and generates API compatible user config structure.
+// Where:
+//   - c is a user config that we get from Terraform state using d.Get(...)
+//   - b is a blueprint of the user config API request
+//
+// It iterates over the data from the Terraform and converts, if necessary,
+// to the API-compatible format.
+func walkBlueprint(c map[string]interface{}, b map[string]interface{}) map[string]interface{} {
+	m := make(map[string]interface{})
+	for ck, cv := range c {
+		if isUserConfigFieldEmpty(cv) {
+			continue
+		}
+
+		for bk, bv := range b {
+			// if key from the user_config is not equal to blueprint skip it
+			if ck != bk {
+				continue
+			}
+
+			if bv == "primitive" { // primitive type
```

Comment:
Since it doesn't really matter what value is,
I would check it for the string type, and write expected type, i.e. float, int, etc.
That would help for debugging.